src/mgmt/srv_query.c

```diff
@@ -0,0 +1,192 @@
+	 */
+	if (!uuid_is_null(dev_uuid)) {
+		rc = smd_dev_get_by_id(dev_uuid, &mbh->mb_dev_info);
+		if (rc != 0) {
+			D_ERROR("Device UUID:"DF_UUID" not found\n",
+				DP_UUID(dev_uuid));
+			return rc;
+		}
+		if (mbh->mb_dev_info->sdi_tgts == NULL) {
+			D_ERROR("No targets mapped to device\n");
+			return -DER_NONEXIST;
+		}
+		/* Default tgt_id is the first mapped tgt */
+		tgt_id = mbh->mb_dev_info->sdi_tgts[0];
+	} else {
+		tgt_id = atoi(tgt);
+		rc = smd_dev_get_by_tgt(tgt_id, &mbh->mb_dev_info);
+		if (rc != 0) {
+			D_ERROR("Tgt_id:%d not found\n", tgt_id);
+			return rc;
+		}
+		uuid_copy(dev_uuid, mbh->mb_dev_info->sdi_id);
+	}
+
+	D_DEBUG(DB_MGMT, "Querying BIO Health Data for dev:"DF_UUID"\n",
+		DP_UUID(dev_uuid));
+	uuid_copy(mbh->mb_devid, dev_uuid);
+
+	/* Create a ULT on the tgt_id */
+	D_DEBUG(DB_MGMT, "Starting ULT on tgt_id:%d\n", tgt_id);
+	rc = dss_ult_create(bio_health_query, mbh, DSS_ULT_BIO, tgt_id, 0,
+			    &thread);
+	if (rc != 0) {
+		D_ERROR("Unable to create a ULT on tgt_id:%d\n", tgt_id);
+		return rc;
+	}
+
+	ABT_thread_join(thread);
+	ABT_thread_free(&thread);
+
+	return rc;
+}
+
+int
+ds_mgmt_smd_list_devs(Mgmt__SmdDevResp *resp)
+{
+	struct smd_dev_info	*dev_info, *tmp;
+	d_list_t		 dev_list;
+	int			 dev_list_cnt = 0;
+	int			 i = 0, j;
+	int			 rc = 0;
+
+	D_DEBUG(DB_MGMT, "Querying SMD device list\n");
+
+	D_INIT_LIST_HEAD(&dev_list);
+	rc = smd_dev_list(&dev_list, &dev_list_cnt);
+
+	D_ALLOC(resp->devices, sizeof(*resp->devices) * dev_list_cnt);
+	if (resp->devices == NULL) {
+		D_ERROR("Failed to allocate devices for resp\n");
+		return -DER_NOMEM;
+	}
+
+	d_list_for_each_entry_safe(dev_info, tmp, &dev_list, sdi_link) {
+		D_ALLOC(resp->devices[i], sizeof(*(resp->devices[i])));
+		if (resp->devices[i] == NULL) {
+			D_ERROR("Failed to allocate device\n");
+			rc = -DER_NOMEM;
+			break;
+		}
+		mgmt__smd_dev_resp__device__init(resp->devices[i]);
+		D_ALLOC(resp->devices[i]->uuid, DAOS_UUID_STR_SIZE);
+		if (resp->devices[i]->uuid == NULL) {
+			D_ERROR("Failed to allocate device uuid\n");
+			rc = -DER_NOMEM;
+			break;
+		}
+		uuid_unparse_lower(dev_info->sdi_id, resp->devices[i]->uuid);
+
+		resp->devices[i]->n_tgt_ids = dev_info->sdi_tgt_cnt;
+		D_ALLOC(resp->devices[i]->tgt_ids,
+			sizeof(int) * dev_info->sdi_tgt_cnt);
+		if (resp->devices[i]->tgt_ids == NULL) {
+			rc = -DER_NOMEM;
+			break;
+		}
+		for (j = 0; j < dev_info->sdi_tgt_cnt; j++)
+			resp->devices[i]->tgt_ids[j] = dev_info->sdi_tgts[j];
+
+		d_list_del(&dev_info->sdi_link);
+		/* Frees sdi_tgts and dev_info */
+		smd_free_dev_info(dev_info);
+
+		i++;
+	}
+	/* Free all devices is there was an error allocating any */
+	if (rc != 0) {
+		for( ; i >= 0; i-- ) {
+			if (resp->devices[i] != NULL) {
+				if (resp->devices[i]->uuid != NULL)
+					D_FREE(resp->devices[i]->uuid);
+				if (resp->devices[i]->tgt_ids != NULL)
+					D_FREE(resp->devices[i]->tgt_ids);
+				D_FREE(resp->devices[i]);
```

Comment:
Can we do all of these in one allocation/free? It looks to me they are part of dRPC reply buffer?